rali/rali/source/tf_record_reader.cpp

```diff
@@ -174,106 +187,113 @@ void TFRecordReader::replicate_last_image_to_fill_last_shard()
+        uint32_t length_crc, data_crc;
+
         length = file_contents.tellg();
-        char *header_length = new char[uint64_size];
-        char *header_crc = new char[uint32_size];
-        char *footer_crc = new char[uint32_size];
-        file_contents.read(header_length, uint64_size);
+        file_contents.read((char *)&data_length, sizeof(data_length));
         if(!file_contents)
             THROW("TFRecordReader: Error in reading TF records")
-        file_contents.read(header_crc, uint32_size);
+        file_contents.read((char *)&length_crc, sizeof(length_crc));
         if(!file_contents)
             THROW("TFRecordReader: Error in reading TF records")
-        memcpy(&data_length, header_length, sizeof(data_length));
-        memcpy(&length_crc, header_crc, sizeof(length_crc));
-        if(uint(length + data_length + 16) == file_size)
+        if (uint(length + data_length + 16) == file_size)
         {
             _last_rec = true;
-
         }
-        char *data = new char[data_length];
-        file_contents.read(data,data_length);
+        std::unique_ptr<char[]> data(new char[data_length]);
+        file_contents.read(data.get(), data_length);
         if(!file_contents)
-            THROW("TFRecordReader: Error in reading TF records")
-        _single_example.ParseFromArray(data,data_length);
+            THROW("TFRecordReader: Error in reading TF records")  
+        _single_example.ParseFromArray(data.get(), data_length);
         _features = _single_example.features();
         auto feature = _features.feature();
-        _single_feature = feature.at(_filename_key);
-        std::string fname = _single_feature.bytes_list().value()[0];
+        std::string file_path = _folder_path;
+        std::string fname;
+        if (!_filename_key.empty()) {
+            _single_feature = feature.at(_filename_key);
+            fname = _single_feature.bytes_list().value()[0];
+            file_path.append("/");
+            file_path.append(fname);
+        } else {
+            // generate filename based on file_id
+            fname = std::to_string(_file_id);
+            file_path.append("/");
+            file_path.append(fname);
+        }
         _image_record_starting.insert(std::pair<std::string, uint>(fname, length));
         _in_batch_read_count++;
-        _in_batch_read_count = (_in_batch_read_count%_batch_count == 0) ? 0 : _in_batch_read_count;
-        std::string file_path = _folder_path;
-        file_path.append("/");
-        file_path.append(fname);
+        _in_batch_read_count = (_in_batch_read_count % _batch_count == 0) ? 0 : _in_batch_read_count;
         _last_file_name = file_path;
-        if(get_file_shard_id() != _shard_id )
+        if (get_file_shard_id() != _shard_id)
         {
             incremenet_file_id();
-            file_contents.read(footer_crc, sizeof(data_crc));
+            _file_count_all_shards++;
+            file_contents.read((char *)&data_crc, sizeof(data_crc));
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=6366586)